cpp/src/parquet/stream_writer.h

```diff
@@ -0,0 +1,187 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_STREAM_WRITER_H
+#define PARQUET_STREAM_WRITER_H
+
+#include <array>
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/util/string_view.h"
+#include "parquet/column_writer.h"
+#include "parquet/file_writer.h"
+
+namespace parquet {
+
+/// \brief A class for writing Parquet files using an output stream type API.
+///
+/// The values given must be of the correct type i.e. the type must
+/// match the file schema exactly otherwise a ParquetException will be
+/// thrown.
+///
+/// The user must explicitly indicate the end of the row using the
+/// EndRow() function or EndRow output manipulator.
+///
+/// A maximum row group size can be configured, the default size is
+/// 512MB.  Alternatively the row group size can be set to zero and the
+/// user can create new row groups by calling the EndRowGroup()
+/// function or using the EndRowGroup output manipulator.
+///
+/// Currently there is no support for optional or repeated fields.
+///
+class PARQUET_EXPORT StreamWriter {
+ public:
+  // N.B. Default constructed objects are not usable.  This
+  //      constructor is provided so that the object may be move
+  //      assigned afterwards.
+  StreamWriter() = default;
+
+  explicit StreamWriter(std::unique_ptr<ParquetFileWriter> writer);
+
+  ~StreamWriter();
+
+  static void SetDefaultMaxRowGroupSize(int64_t max_size);
+
+  void SetMaxRowGroupSize(int64_t max_size);
+
+  // Moving is possible.
+  StreamWriter(StreamWriter&&) noexcept = default;
+  StreamWriter& operator=(StreamWriter&&) noexcept = default;
```

Comment:
This causes build failure with g++ 4.8.5 on CentOS 7: https://issues.apache.org/jira/browse/PARQUET-1688

It seems that g++ 4.8.5 doesn't have the default implementation of `operator=() noexcept`:

```text
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.h:31:0,
                 from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.cc:18:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.h:67:17: error: function 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)'
   StreamWriter& operator=(StreamWriter&&) noexcept = default;
                 ^
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.cc:18:0:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.h:61:17: error: function 'parquet::StreamReader& parquet::StreamReader::operator=(parquet::StreamReader&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamReader& parquet::StreamReader::operator=(parquet::StreamReader&&)'
   StreamReader& operator=(StreamReader&&) noexcept = default;
                 ^
make[2]: *** [src/parquet/CMakeFiles/parquet_objlib.dir/stream_reader.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.cc:18:0:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.h:67:17: error: function 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)'
   StreamWriter& operator=(StreamWriter&&) noexcept = default;
                 ^
```